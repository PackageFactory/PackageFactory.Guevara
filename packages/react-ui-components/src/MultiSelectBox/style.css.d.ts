export interface IStyleCss {
  'selectedOptions': string;
}

export const locals: IStyleCss;
