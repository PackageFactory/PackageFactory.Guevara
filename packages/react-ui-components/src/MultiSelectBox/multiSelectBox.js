import React, {PureComponent, PropTypes} from 'react';
import AbstractSelectBox, {propTypes as abstractSelectBoxPropTypes, state as abstractState} from '../SelectBox/abstractSelectBox';

export default class MultiSelectBox extends AbstractSelectBox {
    static propTypes = {
        ...abstractSelectBoxPropTypes,

        value: PropTypes.arrayOf(
            PropTypes.string
        ),

        /**
         * An optional css theme to be injected.
         */
        theme: PropTypes.shape({/* eslint-disable quote-props */
            'selectedOptions': PropTypes.string,
            'selectedOptions__item': PropTypes.string
        }).isRequired, /* eslint-enable quote-props */

        //
        // Static component dependencies which are injected from the outside (index.js)
        // Used in sub-components
        //
        SelectBoxComponent: PropTypes.any.isRequired,
        IconComponent: PropTypes.any.isRequired,
        IconButtonComponent: PropTypes.any.isRequired
    };

    state = {
        ...abstractState,
        selectedOptions: []
    };

    constructor(...args) {
        super(...args);

        this.renderSelectedOption = this.renderSelectedOption.bind(this);
        this.addValueToSelectedOptions = this.addValueToSelectedOptions.bind(this);
        this.handleRemoveSelectedOptionsClick = this.handleRemoveSelectedOptionsClick.bind(this);
    }

    componentDidMount() {
        if (this.props.value) {
            const selectedOptions = [];
            this.props.value.forEach(value => {
                selectedOptions.push({
                    value: value,
                    label: this.getOptionLabelForValue(value),
                    icon: this.getOptionIconForValue(value)
                });
            });

            this.setState({
                selectedOptions: selectedOptions
            });
        }
    }

    render() {
        const {
            options,
            loadOptionsOnInput,
            placeholder,
            placeholderIcon,
            theme,
            minimumResultsForSearch,
            SelectBoxComponent
        } = this.props;

        const selectedOptions = this.state.selectedOptions;

        return (
            <div className={this.props.theme.wrapper}>
                {Object.prototype.toString.call(selectedOptions) === '[object Array]' ?
                    <ul className={this.props.theme.selectedOptions}>
                        {
                            selectedOptions.map(this.renderSelectedOption)
                        }
                    </ul> : null
                }
                <SelectBoxComponent
                    value={null}
                    options={options}
                    loadOptionsOnInput={loadOptionsOnInput}
                    placeholder={placeholder}
                    placeholderIcon={placeholderIcon}
                    theme={theme}
                    onSelect={this.addValueToSelectedOptions}
                    onDelete={null}
                    clearOnSelect={true}
                    minimumResultsForSearch={minimumResultsForSearch}
                />
            </div>
        );
    }

    addValueToSelectedOptions(value) {
        let currentSelectedOptions = [...this.state.selectedOptions || []];

        const valueAlreadySelected = currentSelectedOptions.find(option => option.value == value);

        // TODO fix this, options in state are empty
        console.log (value, this.state.options, this.props.options);

        if (!valueAlreadySelected) {
            currentSelectedOptions.push({
                icon: this.getOptionIconForValue(value),
                label: this.getOptionLabelForValue(value),
                value: value
            });

            this.setState({
                selectedOptions: currentSelectedOptions
            });

            this.props.onSelect(currentSelectedOptions);
        }

    }

    handleRemoveSelectedOptionsClick(valueToRemove) {
        let currentSelectedOptions = [...this.state.selectedOptions];
        const newSelectedOptions = currentSelectedOptions.filter(option => option.value != valueToRemove);

        this.setState({
            selectedOptions: newSelectedOptions
        });

        this.props.onSelect(currentSelectedOptions);
    }

    select(incomingValue, shouldTriggerOnSelect = true) {
        console.log (incomingValue, shouldTriggerOnSelect)
    }

    /**
     * renders a single option (<li/>) for the list of multi selected values
     *
     * @param {object} option
     * @param {string} option.icon
     * @param {string} option.label
     * @param {number} index
     * @returns {JSX} option element
     */
    renderSelectedOption({icon, label, value}, index) {
        const theme = this.props.theme;
        const IconComponent = this.props.IconComponent;
        const IconButtonComponent = this.props.IconButtonComponent;
        const onClick = () => {
            this.handleRemoveSelectedOptionsClick(value);
        };

        return (
            <li
                key={index}
                className={theme.selectedOptions__item}
            >
                {
                    icon ?
                        <IconComponent className={theme.dropDown__itemIcon} icon={icon}/> :
                        null
                }
                <span>{ label }</span>
                <IconButtonComponent
                    icon={'close'}
                    onClick={onClick}
                />
            </li>
        );
    }
}
