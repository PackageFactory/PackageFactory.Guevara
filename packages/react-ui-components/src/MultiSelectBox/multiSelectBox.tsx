/* eslint-disable camelcase, react/jsx-pascal-case */
import React, {PureComponent, ComponentClass} from 'react';
import {$get} from 'plow-js';
import mergeClassNames from 'classnames';
import {omit} from 'lodash';
import SelectBox_Option_SingleLine from '../SelectBox_Option_SingleLine';
import SelectBox from '../SelectBox';
import {SelectBoxOptions, SelectBoxOptionValueAccessor} from '../SelectBox/selectBox';
import {PickDefaultProps} from '../../types';
import MultiSelectBox_ListPreviewSortable from '../MultiSelectBox_ListPreviewSortable';
import {SelectBox_Option_SingleLineProps} from '../SelectBox_Option_SingleLine/selectBox_Option_SingleLine';

export interface MultiSelectBoxProps extends Readonly<{
    // ------------------------------
    // Basic Props for core functionality
    // ------------------------------
    /**
     * This prop represents a set of options.
     * Each option must have a value and can have a label and an icon.
     */
    options: SelectBoxOptions;

    /**
     * Additional className wich will be applied
     */
    className?: string;

    /**
     * Field name specifying which field in a single "option" contains the "value"
     */
    optionValueField: string;

    /**
     * This prop represents the current selected value.
     */
    values: ReadonlyArray<string>;

    /**
     * This prop gets called when an option was selected. It returns the new values as array.
     */
    onValuesChange: (values: ReadonlyArray<string>) => void;

    // ------------------------------
    // Visual customization of the MultiSelect Box
    // ------------------------------
    /**
     * This prop is the placeholder text which is displayed in the selectbox when no option was selected.
     */
    placeholder?: string;

    /**
     * This prop is an icon for the placeholder.
     */
    placeholderIcon?: string;

    /**
     * Text for the group label of options without a group
     */
    withoutGroupLabel?: string;

    /**
     * If false, prevents removing the last element.
     */
    allowEmpty?: boolean;

    /**
     * Limit height and show scrollbars if needed, defaults to true
     */
    scrollable: boolean;

    /**
     * Component used for rendering the individual option elements; Usually this component uses "SelectBoxOption" internally for common styling.
     */
    ListPreviewElement?: ComponentClass<SelectBox_Option_SingleLineProps>; // TODO type the Props

    disabled: boolean;

    // ------------------------------
    // Asynchronous loading of data
    // ------------------------------

    /**
     * This prop is the loading text which is displayed in the selectbox when displayLoadingIndicator ist set to true.
     */
    loadingLabel?: string;

    /**
     * Helper for asynchronous loading; should be set to "true" as long as "options" is not yet populated.
     */
    displayLoadingIndicator: boolean;

    // ------------------------------
    // Search-As-You-Type related functionality
    // ------------------------------
    displaySearchBox: boolean;
    searchTerm: string;
    onSearchTermChange: (searchTerm: string) => void;
    searchOptions?: SelectBoxOptions;

    /**
     * If set to true, the search box is directly focussed once the SelectBox is rendered;
     * such that the user can start typing right away.
     */
    setFocus: boolean;

    // ------------------------------
    // "Create new if not exists" functionality
    // ------------------------------
    /**
     * This prop gets called when requested to create a new element
     */
    onCreateNew: () => void; // TODO

    /**
     * "Create new" label
     */
    createNewLabel: string;

    // ------------------------------
    // Drag&Drop Reordering of Selected Values
    // ------------------------------
    /**
     * Specifying the dnd type. Defaults to 'multiselect-box-value'
     */
    dndType?: string;

    // ------------------------------
    // Theme & Dependencies
    // ------------------------------
    theme: MultiSelectBoxTheme;
}> {}

interface MultiSelectBoxTheme extends Readonly<{
    'selectedOptions': string;
    'selectedOptions__item': string;
    'wrapper': string;
}> {}

export const defaultProps: PickDefaultProps<MultiSelectBoxProps, 'optionValueField' | 'dndType' | 'allowEmpty' | 'ListPreviewElement' | 'disabled'> = {
    optionValueField: 'value',
    dndType: 'multiselect-box-value',
    allowEmpty: true,
    ListPreviewElement: SelectBox_Option_SingleLine,
    disabled: false,
};

export default class MultiSelectBox extends PureComponent<MultiSelectBoxProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element {
        const {
            searchOptions,
            values,
            optionValueField,
            theme,
            className,
            disabled
        } = this.props;

        const filteredSearchOptions = (searchOptions || [])
            .filter(option => !(values && values.indexOf(option[optionValueField]) !== -1));

        const selectedOptionsClassNames = mergeClassNames({
            [theme.selectedOptions]: true
        });

        const optionValueAccessor = this.getOptionValueAccessor();

        const classNames = mergeClassNames(className, theme.wrapper);

        return (
            <div className={classNames}>
                <ul className={selectedOptionsClassNames}>
                    <MultiSelectBox_ListPreviewSortable
                        {...omit(this.props, ['theme'])}
                        optionValueAccessor={optionValueAccessor}
                        disabled={disabled}
                    />
                </ul>
                <SelectBox
                    {...omit(this.props, ['theme', 'className'])}
                    options={filteredSearchOptions}
                    value=""
                    onValueChange={this.handleNewValueSelected}
                    disabled={disabled}
                />
            </div>
        );
    }

    private getOptionValueAccessor(): SelectBoxOptionValueAccessor {
        const {optionValueField} = this.props;
        // @ts-ignore
        return $get([optionValueField]);
    }

    private handleNewValueSelected = (value: string) => {
        const {onValuesChange} = this.props;
        const values = this.props.values || [];
        const updatedValues = [...values, value];

        onValuesChange(updatedValues);
    }
}
