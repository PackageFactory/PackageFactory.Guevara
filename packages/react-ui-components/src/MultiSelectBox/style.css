.selectedOptions {
    composes: reset from './../reset.css';

    position: relative;
    width: 100%;
    list-style: none;
    padding-left: 0;

}

.selectedOptions--highlight {
    outline: 2px solid var(--brandColorsSuccess);
    margin-bottom: 2px;
}
