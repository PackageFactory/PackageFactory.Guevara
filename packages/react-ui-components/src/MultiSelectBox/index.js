import {themr} from 'react-css-themr';
import identifiers from './../identifiers.js';
import style from './style.css';
import MultiSelectBox from './multiSelectBox.js';

const ThemedMultiSelectBox = themr(identifiers.multiSelectBox, style)(MultiSelectBox);

//
// Dependency injection
//
import injectProps from './../_lib/injectProps.js';
import SelectBox from './../SelectBox/index';
import Icon from './../Icon/index';
import IconButton from './../IconButton/index';

export default injectProps({
    IconComponent: Icon,
    IconButtonComponent: IconButton,
    SelectBoxComponent: SelectBox
})(ThemedMultiSelectBox);
