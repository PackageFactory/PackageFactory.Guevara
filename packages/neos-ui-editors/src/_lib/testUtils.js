import {PureComponent, Children} from 'react';
import PropTypes from 'prop-types';
import {SynchronousRegistry, SynchronousMetaRegistry} from '@neos-project/neos-ui-extensibility/src/registry';

const globalRegistry = new SynchronousMetaRegistry();

// i18n Registry
class FakeI18NRegistry extends SynchronousRegistry {
    translate(key) {
        return key;
    }
}
globalRegistry.set('i18n', new FakeI18NRegistry());

// Data Source DataLoader
export const MockDataSourceDataLoader = {
    _currentPromise: null,
    _currentPromiseResolveFn: null,

    // API in tests
    reset() {
        this._currentPromise = null;
    },

    resolveCurrentPromise(result) {
        this._currentPromiseResolveFn(result);
        return this._currentPromise;
    },

    // API of DataLoader
    resolveValue() {
        if (!this._currentPromise) {
            this._currentPromise = new Promise(resolve => {
                this._currentPromiseResolveFn = resolve;
            });
        }
        return this._currentPromise;
    }
};
globalRegistry.set('dataLoaders', new SynchronousRegistry());
globalRegistry.get('dataLoaders').set('DataSources', MockDataSourceDataLoader);

// wrapper
export class WrapWithMockGlobalRegistry extends PureComponent {
    static propTypes = {
        children: PropTypes.element.isRequired
    };

    static childContextTypes = {
        globalRegistry: PropTypes.object.isRequired,
        configuration: PropTypes.object.isRequired
    };

    getChildContext() {
        const configuration = {};
        return {globalRegistry, configuration};
    }

    render() {
        return Children.only(this.props.children);
    }
}
