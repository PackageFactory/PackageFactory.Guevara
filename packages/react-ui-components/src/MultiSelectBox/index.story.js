import React from 'react';
import {storiesOf, action} from '@kadira/storybook';
import {withKnobs, text} from '@kadira/storybook-addon-knobs';
import {StoryWrapper} from './../_lib/storyUtils.js';
import MultiSelectBox from './index.js';

const options = [
    {value: 'opt1', label: 'Option 1'},
    {value: 'opt2', label: 'Option 2'},
    {value: 'opt3', label: 'Option 3'}
];

const selectedOptions = [
    'opt1', 'opt2'
];

const loadOptions = ({callback, value, searchTerm}) => {
    if (value) {
        // simple search for async options
        const filteredOptions = options.filter(option => {
            return option.value === value;
        });

        setTimeout(() => (callback(filteredOptions)), 1000);
    } else if (searchTerm) {
        // simple search for async options
        const filteredOptions = options.filter(option => {
            return option.label.toLowerCase().indexOf(searchTerm.toLowerCase()) !== -1;
        });

        setTimeout(() => (callback(filteredOptions)), 1000);
    } else {
        setTimeout(() => (callback(options)), 1000);
    }
};

storiesOf('MultiSelectBox', module)
    .addDecorator(withKnobs)
    .addWithInfo(
        'default',
        () => (
            <StoryWrapper>
                <MultiSelectBox
                    options={options}
                    placeholder={text('Placeholder', 'Select')}
                    placeholderIcon={text('Placeholder icon', 'bookmark')}
                    onSelect={action('onSelect')}
                    />
            </StoryWrapper>
        ),
        {inline: true}
    )
    .addWithInfo(
        'defaultWithSelectedOptions',
        () => (
            <StoryWrapper>
                <MultiSelectBox
                    value={selectedOptions}
                    options={options}
                    placeholder={text('Placeholder', 'Select')}
                    placeholderIcon={text('Placeholder icon', 'bookmark')}
                    onSelect={action('onSelect')}
                    />
            </StoryWrapper>
        ),
        {inline: true}
    )
    .addWithInfo(
        'asyncOptions',
        () => (
            <StoryWrapper>
                <MultiSelectBox
                    options={loadOptions}
                    placeholder={text('Placeholder', 'Select')}
                    placeholderIcon={text('Placeholder icon', 'bookmark')}
                    onSelect={action('onSelect')}
                    />
            </StoryWrapper>
        ),
        {inline: true}
    )
    .addWithInfo(
        'asyncOptionsWithDefault',
        () => (
            <StoryWrapper>
                <MultiSelectBox
                    value={selectedOptions}
                    options={loadOptions}
                    placeholder={text('Placeholder', 'Select')}
                    placeholderIcon={text('Placeholder icon', 'bookmark')}
                    onSelect={action('onSelect')}
                    />
            </StoryWrapper>
        ),
        {inline: true}
    );
