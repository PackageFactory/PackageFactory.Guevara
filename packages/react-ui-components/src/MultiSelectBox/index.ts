/* eslint-disable camelcase, react/jsx-pascal-case */
import {themr} from '@friendsofreactjs/react-css-themr';
import identifiers from '../identifiers';
import style from './style.css';
import MultiSelectBox from './multiSelectBox';

const ThemedMultiSelectBox = themr(identifiers.multiSelectBox, style)(MultiSelectBox);

export default ThemedMultiSelectBox;
