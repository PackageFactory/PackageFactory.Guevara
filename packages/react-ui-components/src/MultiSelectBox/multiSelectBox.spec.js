import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {createStubComponent} from './../_lib/testUtils.js';
import MultiSelectBox from './multiSelectBox.js';

describe('<MultiSelectBox/>', () => {
    let props;

    beforeEach(() => {
        props = {
            onValuesChange: jest.fn(),
            theme: {},
            values: [],
            SelectBoxComponent: createStubComponent(),
            IconComponent: createStubComponent(),
            IconButtonComponent: createStubComponent()
        };
    });

    it('should render correctly.', () => {
        const wrapper = shallow(<MultiSelectBox {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
