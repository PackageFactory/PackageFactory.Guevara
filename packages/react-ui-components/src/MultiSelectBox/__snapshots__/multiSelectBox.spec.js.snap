// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSelectBox/> should render correctly. 1`] = `
<div>
  <ul
    className=""
  >
    <Component
      IconButtonComponent={[Function]}
      IconComponent={[Function]}
      ListPreviewElement={[Function]}
      MultiSelectBox_ListPreviewSortable={[Function]}
      SelectBox={[Function]}
      allowEmpty={true}
      dndType="multiselect-box-value"
      onCreateNew={[Function]}
      onValuesChange={[Function]}
      optionValueAccessor={[Function]}
      optionValueField="value"
      values={Array []}
    />
  </ul>
  <Component
    IconButtonComponent={[Function]}
    IconComponent={[Function]}
    ListPreviewElement={[Function]}
    MultiSelectBox_ListPreviewSortable={[Function]}
    SelectBox={[Function]}
    allowEmpty={true}
    dndType="multiselect-box-value"
    highlight={false}
    onCreateNew={[Function]}
    onValueChange={[Function]}
    onValuesChange={[Function]}
    optionValueField="value"
    options={Array []}
    value=""
    values={Array []}
  />
</div>
`;
